### PR TITLE
Add IGate rules to standard equivalence library

### LINKED ISSUE #9485: `transpile` function with `basis_gates=['u', 'cx']` fails with any circuit with an `id` gate
### Environment

- **Qiskit Terra version**: 0.22.3
- **Python version**: 3.9.15

### What is happening?

Suppose I am given an arbitrary quantum circuit and I want to transpile it by using only the gate $U(\theta, \phi, \lambda)$ (single-qubit rotation) and the gate $CX$ (creating entanglement), knowing that they form a universal set of gates for quantum computation.
In Qiskit this should be possible calling the `transpile` function and passing the argument `basis_gates=['u', 'cx']`. This works fine in almost any case but fails for every circuit with an identity `id` gate (or more than one).

### How can we reproduce the issue?

```
from qiskit import QuantumCircuit, transpile

qc = QuantumCircuit(1)
qc.id(0)
tqc = transpile(qc, basis_gates=['u', 'cx'])
```

`QiskitError: "Cannot unroll the circuit to the given basis, ['u', 'cx']. Instruction id not found in equivalence library and no rule found to expand."`

### What should happen?

I would expect the transpiler not to raise any error and to complete the circuit transpilation correctly.

### Any suggestions?

To solve this problem, it should be enough to add $I = U(0,0,0)$ to the Qiskit equivalence library, so that the transpiler runs successfully and returns the circuit
ugate

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit adds a missing rules to the equivalence library for an
identity gate. Without this the basis translator has difficulty finding
a match if the basis does not include an id gate (see
Qiskit/qiskit-metapackage#1346)

### Details and comments

Fix #9485.

### CODE REVIEW

**Review by jakelishman — DISMISSED** [review]

**Comment by jakelishman** [comment]
Yeah, I'd seen the synthetic benchmark.  I suppose it makes sense that the basis-gate translator would be run first; if we ran the hypothetical "check for no-ops" too early, we'd probably have to run it again later to make sure we hadn't introduced any more by cancellations or parameter combinations or something.

**Review by kdk — CHANGES_REQUESTED** [review]
Comments in https://github.com/Qiskit/qiskit-metapackage/pull/1346#issuecomment-944373753 .

**Comment by mtreinish** [comment]
> Comments in [Qiskit/qiskit#1346 (comment)](https://github.com/Qiskit/qiskit/pull/1346#issuecomment-944373753) .

This shouldn't hold true anymore as new backends will no longer support the `id` instruction. It's also been raising a warning for > a year if a user submits a circuit to the ibmq provider with an id gate right?

**Comment by mtreinish** [comment]
Additionally I just pushed: https://github.com/Qiskit/qiskit-ibm-provider/pull/468 to have IBM backend's preserve their custom definition of id == delay(sx.duration) through transpile  via custom backend plugin defaults. This means that combined with this PR IBM backends get their custom behavior and for all other backends id == identity (especially when combined with #7403 )

**Review by jakelishman — APPROVED** [review]
Do we need to supply equivalences to other no-op gates, or do you think it would be suitable to just supply and equivalence to the empty 1q DAG?

If we do need the no-op gates, then this seems fine to me.

File: qiskit/circuit/library/standard_gates/equivalence_library.py
Comment (kdk) [inline qiskit/circuit/library/standard_gates/equivalence_library.py:183]:
```suggestion
q = QuantumRegister(1, "q")
def_id = QuantumCircuit(q)
_sel.add_equivalence(IGate(), def_id)
```

I think this is the same as @jakelishman 's question, but is there a reason to replace `IGate` with other effective nop gates as opposed to replacing with an empty circuit? If we had a subsequent pass in optimization to remove nops, then I think it would be easier to leave `IGate`'s in place here (and add `'id'` to the `BasisTranslator`'s `basic_instrs`) and find them later, as opposed to having to inspect gates in optimization to see which are effective no-ops (which we do anyway in synthesis, but is computationally harder).

I suppose one argument against this is that removing `IGate`s here might be considered an optimization, which we wouldn't want in the `BasisTranslator` or in e.g. `optimization_level=0`, but that `IGate`s also aren't supported natively on all devices. However, I think this requires some thought as not all of these nop implementations will have the same cost/behavior on devices. e.g. `UGate(0,0,0)` will on IBM devices be implemented with two single qubit `sx` gates, while `RZGate(0)` will require none.

**Comment by kdk** [comment]
> Additionally I just pushed: [Qiskit/qiskit-ibm-provider#468](https://github.com/Qiskit/qiskit-ibm-provider/pull/468) to have IBM backend's preserve their custom definition of id == delay(sx.duration) through transpile via custom backend plugin defaults. This means that combined with this PR IBM backends get their custom behavior and for all other backends id == identity (especially when combined with #7403 )

I'm a bit confused as to why https://github.com/Qiskit/qiskit-ibm-provider/pull/468 is needed in addition to the this PR and #7403. Wouldn't that override this PR and #7403 to reinstate the deprecated behavior (and maybe silence the deprecation warning)?

**Comment by jakelishman** [comment]
Yeah, that's the idea: the old backends that _do_ maintain that definition translate `id` into the `delay` they associate with it themselves, so the rest of the transpiler stack can move away from it, without affecting the old backends.

This and #7403 are about not enforcing that custom behaviour on every other hardware backend, and letting various other parts of Qiskit now reason about the identity operation at a high-level in the expected way.

**Comment by kdk** [comment]
> Yeah, that's the idea: the old backends that _do_ maintain that definition translate `id` into the `delay` they associate with it themselves, so the rest of the transpiler stack can move away from it, without affecting the old backends.

I think I'm a bit lost on the motivation. It might be a good time to write down all the cases we're thinking of, and the various timelines and expected behaviors. Why would the backends that still maintain `'id'` want or need a conversion into the equivalent `'delay'`? Is the expectation that these backends will maintain an `'id'` until they're retired?

File: qiskit/circuit/library/standard_gates/equivalence_library.py
Comment (jakelishman) [reply qiskit/circuit/library/standard_gates/equivalence_library.py:193]: I'm still in favour of this definition that Kevin's suggesting above, but I also in principle don't hugely mind leaving the definitions that are already in the PR.

Fwiw, the level of basis translation doesn't / shouldn't resolve or know the number of pulses used to implement a device, so I wouldn't call a translation of `IGate` to the empty body an "optimisation", I'd just call it a valid translation to something a backend can run.  If the backend natively supports `IGate`, no translation occurs.  It's up to pulse-aware passes to resynthesise 1q / 2q blocks into pulse-optimal sequences, if the user chooses that.

**Review by jakelishman — APPROVED** [review]
A minor question, but I'm in general happy to merge.

File: test/python/transpiler/test_basis_translator.py
Comment (jakelishman) [inline test/python/transpiler/test_basis_translator.py:938]: My question from the prior review! It got lost before - I must have forgotten to save the comment.

Do we know why this changes?  We haven't added any rules that go _to_ `IGate`, so I'd not have expected the basis-translation path to change.

Comment (mtreinish) [reply test/python/transpiler/test_basis_translator.py:938]: TBH, I'm not sure. I didn't really think to check. I just saw the output circuit was equivalent and updated the test. I'll do some debugging locally because the qpy tests are failing in a weird way now too so this is blocked on that anyway.

File: qiskit/circuit/library/standard_gates/equivalence_library.py
Comment (kdk) [reply qiskit/circuit/library/standard_gates/equivalence_library.py:193]: Following the discussion in https://github.com/Qiskit/qiskit/pull/7403#pullrequestreview-1537269508 , the code as written looks to be the right approach.

**Review by kdk — APPROVED** [review]
Following some updates since this was last discussed (`id` will stick around on IBM backends, and is being defined in Qiskit as a single-qubit no-op of backend-defined duration), I think this is good to go (pending https://github.com/Qiskit/qiskit/pull/7146#discussion_r1321862135 ).

It would also be good to call out (in the `IGate` docstring and maybe as a release note) that `IGate` is no longer strictly opaque, and should be expected to be optimized away by the transpiler except at `optimization_level=0`.

**Pull request merged**
